commit 2bfc5ca48a44
Author: Eric Rahm <erahm@mozilla.com>
Date:   Wed Jan 17 16:38:18 2018 -0800

    Bug 1428612 - Part 1: Use nsLiteralCString to wrap histogram names. r=gfritzsche a=lizzard
    
    This switches from `nsDependentCString` to `nsLiteralCString` when adding
    histograms to the table. It will help save some memory and should avoid the OOM
    smalls we have been seeing.
    
    --HG--
    extra : source : 13493806c0472ebd52ab4140312747c5fbd42fbc
---
 toolkit/components/telemetry/TelemetryHistogram.cpp | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git toolkit/components/telemetry/TelemetryHistogram.cpp toolkit/components/telemetry/TelemetryHistogram.cpp
index a90e5d24aab8..eb28684fe4d1 100644
--- toolkit/components/telemetry/TelemetryHistogram.cpp
+++ toolkit/components/telemetry/TelemetryHistogram.cpp
@@ -1740,9 +1740,14 @@ void TelemetryHistogram::InitializeGlobalState(bool canRecordBase,
   // declaration point further up in this file.
 
   // Populate the static histogram name->id cache.
-  // Note that the histogram names are statically allocated.
+  // Note that the histogram names come from a static table so we can wrap them
+  // in a literal string to avoid allocations when it gets copied.
   for (uint32_t i = 0; i < HistogramCount; i++) {
-    gNameToHistogramIDMap.Put(nsDependentCString(gHistogramInfos[i].name()), HistogramID(i));
+    auto name = gHistogramInfos[i].name();
+
+    nsCString wrappedName;
+    wrappedName.AssignLiteral(name, strlen(name));
+    gNameToHistogramIDMap.Put(wrappedName, HistogramID(i));
   }
 
 #ifdef DEBUG

commit e31ac19b6163
Author: Eric Rahm <erahm@mozilla.com>
Date:   Wed Jan 17 16:52:37 2018 -0800

    Bug 1428612 - Part 2: Add an assertion to check the histogram name pointer. r=gfritzsche a=lizzard
    
    This adds a dignostic assertion that the name pointer is actually within the
    static name table.
    
    --HG--
    extra : source : 6931047d462f46b1f601cb5927e4185857400a63
---
 toolkit/components/telemetry/TelemetryHistogram.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git toolkit/components/telemetry/TelemetryHistogram.cpp toolkit/components/telemetry/TelemetryHistogram.cpp
index eb28684fe4d1..926536e401b4 100644
--- toolkit/components/telemetry/TelemetryHistogram.cpp
+++ toolkit/components/telemetry/TelemetryHistogram.cpp
@@ -1745,6 +1745,11 @@ void TelemetryHistogram::InitializeGlobalState(bool canRecordBase,
   for (uint32_t i = 0; i < HistogramCount; i++) {
     auto name = gHistogramInfos[i].name();
 
+    // Make sure the name pointer is in a valid region. See bug 1428612.
+    MOZ_DIAGNOSTIC_ASSERT(name >= gHistogramStringTable);
+    MOZ_DIAGNOSTIC_ASSERT(
+        uintptr_t(name) < (uintptr_t(gHistogramStringTable) + sizeof(gHistogramStringTable)));
+
     nsCString wrappedName;
     wrappedName.AssignLiteral(name, strlen(name));
     gNameToHistogramIDMap.Put(wrappedName, HistogramID(i));
